### PR TITLE
[stable/dask] Add updated Dask chart

### PR DESCRIPTION
**Pull request by mrocklin**
This creates a new Dask chart that replaces the dask-distributed chart.  In addition to a slimmed down name this uses more modern docker images, supports environment customization through environment variables, links cpu and memory limits to dask cpu and memory limits, ...

See https://github.com/kubernetes/charts/issues/2912

cc @danielfrg

### CODE REVIEW

**Comment by mrocklin** [comment]
Related changes to docker image here: https://github.com/dask/dask-docker/pull/5

**Comment by mrocklin** [comment]
@danielfrg if you have time to review the changes here and in https://github.com/dask/dask-docker/pull/6 I would appreciate it.  

Charts maintainers, I'm curious about how best to go about smoothly renaming a chart.  I would like to rename this from `dask-distributed` to `dask` if possible.

I am also curious about how best to decompose this chart so that it can be used either with or without the Jupyter notebook service.  My current thought is to create a dask-core chart that does not include Jupyter, and then have the Dask chart depend on the dask-core chart.  I am unfamiliar with dependencies though, and so I may be doing something naive here.

I'm running the CLA past the legal department of my company.  I hope for a response back in a day or so.

**Comment by mrocklin** [comment]
@k8s-ci-robot I have signed the CLA

**Comment by danielfrg** [comment]
Changes look good, I cannot test today but I will try to test tomorrow. Maybe lets merge the docker image one to it gets build on dockerhub first.

File: stable/dask/Chart.yaml
Comment (unguiculus) [inline stable/dask/Chart.yaml]: What's the reason for this? You rename the chart and decrease the version? Seems weird.

File: stable/dask/values.yaml
Comment (unguiculus) [inline stable/dask/values.yaml]: Should be a stable version.

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Comment by unguiculus** [comment]
@kubernetes/charts-maintainers The chart is about to be renamed. Is this something we can/want to handle?

**Comment by mrocklin** [comment]
There has been a fair amount of work on this over at https://github.com/dask/helm-chart and things have cleaned up a bit.  If review is restarting let me know and I'll push more up.

File: stable/dask/Chart.yaml
Comment (mrocklin) [reply stable/dask/Chart.yaml]: Happy to increase the version if desired.  I decreased the version to 1.0 because of the name change.  I'm happy to adhere to any standard though.

File: stable/dask/values.yaml
Comment (mrocklin) [reply stable/dask/values.yaml]: Agreed

**Comment by unguiculus** [comment]
@mrocklin Just keep changes coming. It probably doesn't make sense to review now if you push more substantial changes.

**Comment by mrocklin** [comment]
cc @jacobtomlinson I'd like to make you aware of this.  After merging current efforts you may want to consider pushing the result upstream to the stable repo here.

**Comment by unguiculus** [comment]
FYI: Renaming the chart is not as easy. We'd have to keep the existing one and deprecate it. Think about if you really want this.

**Comment by mrocklin** [comment]
Thanks for the comment @unguiculus 

Deprecating and creating a new one still sounds ideal to me.  I may not be aware of maintenance costs of keeping around a deprecated chart though.

**Comment by prydonius** [comment]
Once the chart is deprecated there are no maintenance costs. Be aware that Helm doesn't really treat deprecated charts in any special way and e.g. a `helm search dask` will show both charts. When deprecating dask-distributed, we can update the description to begin with "DEPRECATED" so it's more obvious to users.

We need a separate PR to deprecate dask-distributed that:
- sets `deprecated: true` in Chart.yaml
- prepends "DEPRECATED" to the description in Chart.yaml
- Updates the README and NOTES.txt to indicate that it's deprecated and `stable/dask` should be used instead

This PR should be updated to just add the new dask chart and not remove dask-distributed.

@danielfrg are you happy with renaming this?

**Comment by danielfrg** [comment]
Yes

**Comment by mrocklin** [comment]
I've updated this PR to add a new `dask` chart.  I'll submit a PR deprecating dask-distributed after (if) this gets merged.  Feedback is generally quite welcome on the structure of the chart here.

File: stable/dask/Chart.yaml
Comment (mrocklin) [inline stable/dask/Chart.yaml]: Currently I'm pinning the version of the chart to the version of the main software package.  I'm also happy to go directly to 1.0.0 and follow semver if preferred

Comment (unguiculus) [reply stable/dask/Chart.yaml]: Add `appVersion` which should be `0.17.0`. The chart is versioned independently.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Please read our review guidelines again:
https://github.com/kubenetes/charts/blob/master/REVIEW_GUIDELINES.md

Make sure you configure things as suggested there (e. g. images, resources).

If you use `helm create` with Helm 2.8+ you'll get an initial setup with current best practices applied.

**Comment by unguiculus** [comment]
For the new chart, please compare with the templates and the `_helpers.tpl` file you get with `helm create` using Helm 2.8+ and update accordingly.

**Comment by mrocklin** [comment]
> For the new chart, please compare with the templates and the _helpers.tpl file you get with helm create using Helm 2.8+ and update accordingly.

Thanks for the pointer.  I've half-completed this by adopting the new truncation and stripping conventions.  However I haven't yet adopted the fullnameOverride or `contains $name .Release.Name` logic.  This chart doesn't currently use the `"name.fullname"` construct.  Should I add this anyway?  Is this important for interoperability with other systems?

**Comment by mrocklin** [comment]
The linting checks on circle-ci seem off.  They're testing against the old chart `dask-distributed` (against which this PR was originally targetting) rather than the new chart, `dask`.

**Comment by unguiculus** [comment]
The linter is correct. It tested both charts in this PR. The old one didn't have a version bump. You should really only have the new chart in this PR. Why is there a change to the old one as well that is supposed to be deprecated in another PR?

**Comment by mrocklin** [comment]
> The linter is correct. It tested both charts in this PR. The old one didn't have a version bump. You should really only have the new chart in this PR. Why is there a change to the old one as well that is supposed to be deprecated in another PR?

Ah, thank you.  I missed that change in the diff.

**Comment by mrocklin** [comment]
I could use direction on [this comment](https://github.com/kubernetes/charts/pull/2914#issuecomment-368266148).  Otherwise all concerns have, I believe, been dealt with.

**Review by unguiculus — CHANGES_REQUESTED** [review]
* The empty `stable/dask/NOTES.txt` should be removed
* Please check what `helm create` using Helm 2.8+ gives you and update accordingly
* Make sure you follow our review guidelines

**Comment by mrocklin** [comment]
> The empty stable/dask/NOTES.txt should be removed

Done

> Please check what helm create using Helm 2.8+ gives you and update accordingly

I believe that I have done this, with the exception of the concern raised in https://github.com/kubernetes/charts/pull/2914#issuecomment-368266148 for which I could use help.  Apparently I've failed at this though.  More concrete feedback would be welcome.  

> Make sure you follow our review guidelines

I'm sorry, I looked through these when you made the same comment earlier and I've just read through them again.  I don't see what I'm doing wrong here.  I must be dense here.  If anyone sees what I'm doing wrong I would love detailed feedback.

File: stable/dask/templates/_helpers.tpl
Comment (unguiculus) [inline stable/dask/templates/_helpers.tpl:1]: Please update with latest `_helpers.tpl` you get from `helm create` using Helm 2.8+.

File: stable/dask/README.md
Comment (unguiculus) [inline stable/dask/README.md]: Update image params.

File: stable/dask/templates/dask-jupyter-deployment.yaml
Comment (unguiculus) [inline stable/dask/templates/dask-jupyter-deployment.yaml]: `apps/v1beta2`

File: stable/dask/templates/dask-jupyter-config.yaml
Comment (unguiculus) [inline stable/dask/templates/dask-jupyter-config.yaml]: Once you've updated the `_helpers.tpl`, use `chart: {{ template "dask.chart" . }}`.

File: stable/dask/templates/_helpers.tpl
Comment (unguiculus) [inline stable/dask/templates/_helpers.tpl:8]: Fullname templates should be based on the standard fullname template and append a component suffix.

Is there a reason why you make the components dynamic? I'd just hard-code them as you have them in `values.yaml`.

File: stable/dask/templates/dask-jupyter-deployment.yaml
Comment (unguiculus) [inline stable/dask/templates/dask-jupyter-deployment.yaml]: Make `pullPolicy` configurable.

File: stable/dask/templates/dask-worker-deployment.yaml
Comment (unguiculus) [inline stable/dask/templates/dask-worker-deployment.yaml]: `apps/v1beta2`

File: stable/dask/templates/dask-scheduler-deployment.yaml
Comment (unguiculus) [inline stable/dask/templates/dask-scheduler-deployment.yaml]: Any reason why you make container ports configurable. Changing them may break things. I'd also give them a name and refer to them by name in the services.

File: stable/dask/templates/dask-scheduler-deployment.yaml
Comment (unguiculus) [inline stable/dask/templates/dask-scheduler-deployment.yaml]:
```yaml
          resources: 
{{ toYaml .Values.scheduler.resources | indent 12 }}
```

File: stable/dask/templates/dask-worker-deployment.yaml
Comment (unguiculus) [inline stable/dask/templates/dask-worker-deployment.yaml]:
```yaml
          resources: 
{{ toYaml .Values.scheduler.resources | indent 12 }}
```

File: stable/dask/templates/dask-worker-deployment.yaml
Comment (unguiculus) [inline stable/dask/templates/dask-worker-deployment.yaml]: Just a suggestion for a consistent list style:

```yaml
          args: 
            - dask-worker
            - '{{ template "dask.scheduler-fullname" . }}:{{ .Values.scheduler.servicePort }}'
          {{- if .Values.worker.limits -}}
            - --nthreads
            - {{ default .Values.worker.limits.cpu 1 | quote }}
            - --memory-limit
            - {{ default .Values.worker.limits.memory "4GiB" | quote }}
          {{- end -}}
            - --no-bokeh
```
And don't use the `default` function here. Add the defaults to `values.yaml`.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Sorry, I shouldn't have blindly repeated that. I was on my way home from work and added the comment from my cell phone.

I did a better review now. I'd like you to also indent the lists consistently, either zero or two spaces. Personally, I'd prefer the latter.

**Comment by mrocklin** [comment]
Awesome.  Thank you for your time and for the detailed feedback.

**Comment by mrocklin** [comment]
OK, I've resolved recent issues.  I'm having difficulty testing things locally with minikube.  I'm now getting the following:

```
mrocklin@carbon:~/workspace/charts/stable/dask$ helm install .
Error: release jumpy-heron failed: Deployment in version "v1beta2" cannot be handled as a Deployment: v1beta2.Deployment: Spec: v1beta2.DeploymentSpec: Template: v1.PodTemplateSpec: Spec: v1.PodSpec: Containers: []v1.Container: v1.Container: Image: ReadString: expects " or n, parsing 670 ...,"image":{... at {"apiVersion":"apps/v1beta2","kind":"Deployment","metadata":{"labels":{"app":"dask","chart":"dask-1.0.0","component":"jumpy-heron-jupyter-notebook","heritage":"Tiller","release":"jumpy-heron"},"name":"jumpy-heron-dask-jupyter","namespace":"default"},"spec":{"replicas":1,"strategy":{"type":"RollingUpdate"},"template":{"metadata":{"labels":{"app":"dask","component":"jumpy-heron-jupyter-notebook","release":"jumpy-heron"}},"spec":{"containers":[{"env":[{"name":"EXTRA_CONDA_PACKAGES","value":null},{"name":"EXTRA_PIP_PACKAGES","value":null},{"name":"EXTRA_APT_PACKAGES","value":null},{"name":"DASK_SCHEDULER_ADDRESS","value":"jumpy-heron-dask-scheduler:8786"}],"image":{"pullPolicy":"IfNotPresent","repository":"daskdev/dask-notebook","tag":"0.17.1"},"name":"jumpy-heron-dask-jupyter","ports":[{"containerPort":8888}],"resources":{},"volumeMounts":[{"mountPath":"/home/jovyan/.jupyter","name":"config-volume"}]}],"volumes":[{"configMap":{"name":"jumpy-heron-dask-scheduler-config"},"name":"config-volume"}]}}}}
```

I suspect that this has to do with my local environment

```
mrocklin@carbon:~/workspace/charts/stable/dask$ helm version
Client: &version.Version{SemVer:"v2.8.1", GitCommit:"6af75a8fd72e2aa18a2b278cfe5c7a1c5feca7f2", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.8.1", GitCommit:"6af75a8fd72e2aa18a2b278cfe5c7a1c5feca7f2", GitTreeState:"clean"}
mrocklin@carbon:~/workspace/charts/stable/dask$ kubectl version
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.0", GitCommit:"925c127ec6b946659ad0fd596fa959be43f0cc05", GitTreeState:"clean", BuildDate:"2017-12-15T21:07:38Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"", Minor:"", GitVersion:"v1.9.0", GitCommit:"925c127ec6b946659ad0fd596fa959be43f0cc05", GitTreeState:"clean", BuildDate:"2018-01-26T19:04:38Z", GoVersion:"go1.9.1", Compiler:"gc", Platform:"linux/amd64"}
```

But I haven't yet found a way to resolve this.

File: stable/dask/values.yaml
Comment (mrocklin) [inline stable/dask/values.yaml]: Should we include these?  These get placed into environment variables that trigger certain actions when using the docker images described above.  If someone were to use their own docker image (that didn't inherit from ours) then these options would have no effect.  

We could keep these for convenience or remove them and resolve this through documentation, telling users how to specify environment varibles.

**Comment by mrocklin** [comment]
cc @jacobtomlinson who expressed some interest in this

File: stable/dask/values.yaml
Comment (unguiculus) [reply stable/dask/values.yaml]: I was wondering about those. Since they are empty by default anyways, you should probably remove them and add a generic way to add env vars.

Comment (mrocklin) [reply stable/dask/values.yaml]: Done.  However I've run into a problem where I want to combine environment variables defined within the template and those that are user provided.  Naively I want the following:

```yaml
          env:
            - name: DASK_SCHEDULER_ADDRESS
              value: {{ template "dask.fullname" . }}-scheduler:{{ .Values.scheduler.servicePort }}
{{ toYaml .Values.jupyter.env | indent 12 }}
```

But this generates the following linting error:

```
$ helm lint .
==> Linting .
[ERROR] templates/dask-jupyter-deployment.yaml: unable to parse YAML
	error converting YAML to JSON: yaml: line 40: could not find expected ':'

Error: 1 chart(s) linted, 1 chart(s) failed
```

Comment (mrocklin) [reply stable/dask/values.yaml]: I suspect that there is a clever way to do with with the helm templating language.  Happy to go through those docs again if that's the right approach.

**Comment by jacobtomlinson** [comment]
A few thoughts from me:
- I would like to see an ingress which is disabled by default but I could easily enable and configure to expose the dashboard externally.
- I would like to be able to disable the notebook that comes with it as it is not 100% necessary.
- I would love an option for the cluster to be adaptive. This would require RBAC for the scheduler to modify the worker deployment. It would also be good to have a min and max workers option in distributed as we have previously discussed. (I appreciate this may be out of scope for now though).

**Comment by mrocklin** [comment]
> I would like to see an ingress which is disabled by default but I could easily enable and configure to expose the dashboard externally.

> I would like to be able to disable the notebook that comes with it as it is not 100% necessary.

I'm on board with both of these but have to admit that my familiarity with Helm is limited enough that it's not clear to me today how to implement them.  Do you have any interest in collaborating on this?  I'm giving you push access to my fork now in case the answer is "yes".

> I would love an option for the cluster to be adaptive. This would require RBAC for the scheduler to modify the worker deployment. It would also be good to have a min and max workers option in distributed as we have previously discussed. (I appreciate this may be out of scope for now though).

I agree with this.  If we want min/max then we'll have to wait a bit for that to land in the dask.distributed library.  Things could be incremental though.  Again, I'm not familiar enough with RBAC to be particularly efficient on this.  If you're at all interested I would encourage your involvement.

**Comment by mrocklin** [comment]
OK, I'm currently a bit stuck on a couple things:

1. Despite linting fine my my local environment doesn't run this well.  I don't understand the error messages it's giving me, nor have I yet been able to find suitable information from web searches.

        mrocklin@carbon:~/workspace/charts/stable/dask$ helm install .
        Error: release jumpy-heron failed: Deployment in version "v1beta2" cannot be handled as a Deployment: v1beta2.Deployment: Spec: v1beta2.DeploymentSpec: Template: v1.PodTemplateSpec: Spec: v1.PodSpec: Containers: []v1.Container: v1.Container: Image: ReadString: expects " or n, parsing 670 ...,"image":{... at {"apiVersion":"apps/v1beta2","kind":"Deployment","metadata":{"labels":{"app":"dask","chart":"dask-1.0.0","component":"jumpy-heron-jupyter-notebook","heritage":"Tiller","release":"jumpy-heron"},"name":"jumpy-heron-dask-jupyter","namespace":"default"},"spec":{"replicas":1,"strategy":{"type":"RollingUpdate"},"template":{"metadata":{"labels":{"app":"dask","component":"jumpy-heron-jupyter-notebook","release":"jumpy-heron"}},"spec":{"containers":[{"env":[{"name":"EXTRA_CONDA_PACKAGES","value":null},{"name":"EXTRA_PIP_PACKAGES","value":null},{"name":"EXTRA_APT_PACKAGES","value":null},{"name":"DASK_SCHEDULER_ADDRESS","value":"jumpy-heron-dask-scheduler:8786"}],"image":{"pullPolicy":"IfNotPresent","repository":"daskdev/dask-notebook","tag":"0.17.1"},"name":"jumpy-heron-dask-jupyter","ports":[{"containerPort":8888}],"resources":{},"volumeMounts":[{"mountPath":"/home/jovyan/.jupyter","name":"config-volume"}]}],"volumes":[{"configMap":{"name":"jumpy-heron-dask-scheduler-config"},"name":"config-volume"}]}}}}

2.  I want to drop in a template containing a map in the midst of an ongoing map

    ```yaml
              env:
                - name: DASK_SCHEDULER_ADDRESS
                   value: {{ template "dask.fullname" . }}-scheduler:{{ .Values.scheduler.servicePort }}
    {{ toYaml .Values.jupyter.env | indent 12 }}
    ```

If anyone has any time I would value advice on either of these particular issues.

There are also the issues raised by @jacobtomlinson .  I would love to find a way to turn off the Jupyter server.  Does anyone have any suggestions or examples of completely turning off a segment of the chart completely?

**Comment by jacobtomlinson** [comment]
In terms of turning off segments see the [rbac section](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/master/jupyterhub/templates/hub/rbac.yaml) in zero to jupyterhub as an example. The whole template is wrapped in an if statement that just checks a boolean from the config file.

**Comment by mrocklin** [comment]
Ah, that seems simple enough.  Done.

File: stable/dask/templates/dask-jupyter.yaml
Comment (unguiculus) [inline stable/dask/templates/dask-jupyter.yaml]: Please create separate files per resource.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/dask/templates/dask-jupyter.yaml
Comment (mrocklin) [reply stable/dask/templates/dask-jupyter.yaml]: Fixed

**Comment by mrocklin** [comment]
I remain stuck on these two issues: https://github.com/kubernetes/charts/pull/2914#issuecomment-370036473

I've raised one as a question upstream: https://github.com/kubernetes/helm/issues/3602

If anyone has thoughts on how to resolve these I would welcome advice, pointers to documentation etc..

If people have looked at these and genuinely don't know if there is a solution then that would be good feedback as well. Alternatives would be welcome.

**Comment by unguiculus** [comment]
I'll try and have a look.

**Comment by mrocklin** [comment]
*Thank you* for interceding here @unguiculus .  This works flawlessly for me locally.

Looking at your changes, was the issue that calling toYaml on an empty mapping was generating something like `{}` rather than empty lines?

**Comment by mrocklin** [comment]
I'm going to take another pass over the README and NOTES.  Is there anything else that's necessary here for a 1.0.0 release?

**Comment by mrocklin** [comment]
Just checking in here.  I believe all feedback so far has been handled (except for some of @jacobtomlinson's comments, which I think can probably wait for future enhancements).  Is there anything else I should be doing here to move this forward?

also cc @yuvipanda in case he wants to weigh in.

File: stable/dask/templates/dask-jupyter-config.yaml
Comment (unguiculus) [inline stable/dask/templates/dask-jupyter-config.yaml]: Let me ask again why you make components configurable. Also, they should not be tied to a release. I'd just hard-code them: `jupyter`, `worker`

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/dask/templates/dask-jupyter-config.yaml
Comment (mrocklin) [reply stable/dask/templates/dask-jupyter-config.yaml]: Thanks for the nudge.  Resolved.  My apologies for letting that slip earlier. 

To answer the direct question of why was this here.  My answer is "I don't know".  This PR is a copy-paste-modify job of the current chart in `stable/dask-distributed`, that can hopefully be deprecated once this cleaner version is in.

**Comment by mrocklin** [comment]
Thank you for your patient support everyone.  I'm very glad to see this in.

**Pull request merged**
